### ISSUE TITLE
pubsub: support Nack

### ISSUE THREAD

**Issue by vangent** [google/go-cloud#1455]

**Comment by jba** [comment]
We should support `Nack` as a method on `Message`, as we do `Ack`.

Instead of exposing deadline extension, we should do it behind the scenes. The only thing we need to expose is an option for the maximum time we'll extend the deadline.

**Comment by jba** [comment]
I recommend you look at the `cloud.google.com/go/pubsub` client to see how it does ack deadline extension.

**Comment by vangent** [comment]
SGTM. I was going to do Nack first since it seems simpler.

**Comment by vangent** [comment]
I split deadline extension into #1682 , leaving this issue just for Nack.

**Comment by vangent** [comment]
@jba 

My current plan is:

* Add a `driver.SendNacks` similar to `driver.SendAcks` that drivers will have to implement. I considered modifying the existing signature, to pass two separate slices, but it seems like most drivers need to handle Acks and Nacks differently (e.g., different RPCs), so it's probably simpler to just separate them.

* Add a `driver.Acktion` struct, holding an `AckID` and an `Action` (enum, with Ack and Nack values). We can add an "ExtendDeadline" action later. Feedback on this clever name is welcome.

* Modify `pubsub.Subscription.ackBatcher` to batch `driver.Acktion` instead of `driver.AckID`. The handler will split the slice of `Acktion` into a slice of `AckIDs` for `driver.SendAcks` and a slice of `AckIDs` for `driver.SendNacks`, and call both in parallel (if needed). I considered having a separate batcher for Nacks but that seemed like overkill.

* Add `pubsub.Message.Nack`, similar to the existing `Ack` function. It will always panic if `driver.AckFunc()` returns non-nil.

**Comment by vangent** [comment]
If that sounds good, how do you want the PRs? Options include:

1. One big PR.
2. One PR to add the new driver interface function `SendNacks`.
   2a. With stubs for all providers, then a separate PR per provider to implement.
   2b. With implementations for all providers.
   then a separate PR to add the portable type code and conformance test.

**Comment by jba** [comment]
There is an argument that Nacks should be more frequent (certainly not at the 99th percentile of processing time, if we ever do that). That would suggest a separate batcher with a smaller delay. But I'm OK waiting on that.

Love the name, cute and clever. Now that you've got it out of your system, though, let's pick a name that won't confuse readers (assuming you do stick with a single batcher) . Maybe something generic like `AckInfo`. 

I think one PR is fine. If you prefer to work in pieces, that's OK too. But instead of leaving the test for the end, write it in the first PR and comment out the conformance tests for the unimplemented providers, or have the test accept an Unimplemented error.
